### PR TITLE
[archiveorg] Fix extraction

### LINKED ISSUE #21330: [archive.org] "Unable to extract jwplayer playlist" and a traceback (Not a #7154 dupe)
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2019.05.20**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v', 'https://archive.org/details/cnn200109110848-0929']
[debug] Encodings: locale cp1252, fs mbcs, out cp850, pref cp1252
[debug] youtube-dl version 2019.05.20
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.14393
[debug] exe versions: ffmpeg N-93886-gfbdb3aa179, ffprobe N-93886-gfbdb3aa179
[debug] Proxy map: {}
[archive.org] cnn200109110848-0929: Downloading webpage
ERROR: Unable to extract jwplayer playlist; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpt31j5y3w\build\youtube_dl\YoutubeDL.py", line 796, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpt31j5y3w\build\youtube_dl\extractor\common.py", line 529, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpt31j5y3w\build\youtube_dl\extractor\archiveorg.py", line 45, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpt31j5y3w\build\youtube_dl\extractor\common.py", line 1004, in _search_regex
youtube_dl.utils.RegexNotFoundError: Unable to extract jwplayer playlist; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

## Description

Attempting to download https://archive.org/details/cnn200109110848-0929 . Having the issue shown above. This is not a #7154 dupe as I am not trying to download a collection.

### LINKED ISSUE #23586: Internet Archive: Unable to extract jwplayer playlist
youtube-dl can't download this. Nor can it even find the video on the main page from which I spawned the video window:

    https://blog.archive.org/2019/12/20/lawrence-lessig-being-a-citizen-is-a-public-office-too/

Here's the complete output.

djc

D:\Temp>youtube-dl --verbose **https://archive.org/embed/theydontrepresentus-lessig**
[debug] System config: []
[debug] User config: ['--verbose', '--ignore-errors', '-o', 'G:\\Video\\youtube-dl\\%(title)s - %(id)s.%(ext)s', '--geo-bypass', '-f', 'bestvideo[height<=1080]+bestaudio/best', '--abort-on-unavailable-fragment', '--write-info-json', '--write-description', '--write-annotations', '--write-thumbnail', '--merge-output-format', 'mkv', '--write-sub', '--write-auto-sub', '--sub-format', 'srt/best']
[debug] Custom config: []
[debug] Command-line args: ['--verbose', 'https://archive.org/embed/theydontrepresentus-lessig']
[debug] Encodings: locale cp1252, fs mbcs, out cp1252, pref cp1252
[debug] youtube-dl version 2020.01.01
[debug] Python version 3.4.4 (CPython) - Windows-7-6.1.7601-SP1
[debug] exe versions: ffmpeg git-2019-12-15-ed9279a, ffprobe git-2019-12-15-ed9279a
[debug] Proxy map: {}
[archive.org] theydontrepresentus-lessig: Downloading webpage
ERROR: Unable to extract jwplayer playlist; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpj4acy7ln\build\youtube_dl\YoutubeDL.py", line 796, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpj4acy7ln\build\youtube_dl\extractor\common.py", line 530, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpj4acy7ln\build\youtube_dl\extractor\archiveorg.py", line 45, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpj4acy7ln\build\youtube_dl\extractor\common.py", line 1005, in _search_regex
youtube_dl.utils.RegexNotFoundError: Unable to extract jwplayer playlist; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.

### LINKED ISSUE #23700: [archive.org] Unable to extract jwplayer playlist
## Checklist

- [x] I'm reporting a broken site support issue
- [x] I've verified that I'm running youtube-dl version **2020.01.01**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar bug reports including closed ones
- [x] I've read bugs section in FAQ

## Verbose log

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['https://archive.org/details/beyond_bengal', '-v']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2020.01.01
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.17763
[debug] exe versions: ffmpeg git-2019-12-04-d5274f8, ffprobe git-2019-12-04-d5274f8
[debug] Proxy map: {}
[archive.org] beyond_bengal: Downloading webpage
ERROR: Unable to extract jwplayer playlist; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpj4acy7ln\build\youtube_dl\YoutubeDL.py", line 796, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpj4acy7ln\build\youtube_dl\extractor\common.py", line 530, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpj4acy7ln\build\youtube_dl\extractor\archiveorg.py", line 45, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmpj4acy7ln\build\youtube_dl\extractor\common.py", line 1005, in _search_regex
youtube_dl.utils.RegexNotFoundError: Unable to extract jwplayer playlist; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

## Description

I have tried multiple archive.org video links but to no avail, I even tried passing username & password but no success.
I think "**Unable to extract jwplayer playlist**" is where the issue is

Thank You

### PR DESCRIPTION
**Pull request by TinyToweringTree**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

The archive.org website has changed the way it embeds the JW Player playlist. That's why the regular expression can't find the playlist and the extractor is currently not working.

This pull request updates the extraction of the playlist to work with the new version of archive.org.

There's an `input` element with the class `js-play8-playlist`. Its value contains the playlist.

Closes #21330, closes #23586, closes #23700.

### CODE REVIEW

File: youtube_dl/extractor/archiveorg.py [hunk 1]
```diff
@@ -40,9 +40,12 @@ def _real_extract(self, url):
...
         webpage = self._download_webpage(
             'http://archive.org/embed/' + video_id, video_id)
+        input_element_with_playlist = self._search_regex(
+            r'(<\s*input.*\s*class\s*=\s*[\'"].*\s*js-play8-playlist\s*.*[\'"]\s*.*>)',
+            webpage, 'jwplayer playlist')
         jwplayer_playlist = self._parse_json(self._search_regex(
-            r"(?s)Play\('[^']+'\s*,\s*(\[.+\])\s*,\s*{.*?}\)",
-            webpage, 'jwplayer playlist'), video_id)
+            r'.*\s+value\s*=\s*[\'"](.+)[\'"][\s/]',
+            input_element_with_playlist, 'playlist data'), video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/archiveorg.py]: `extract_attributes`.

File: youtube_dl/extractor/archiveorg.py [hunk 2]
```diff
@@ -52,7 +55,7 @@ def get_optional(metadata, field):
         metadata = self._download_json(
             'http://archive.org/details/' + video_id, video_id, query={
                 'output': 'json',
-            })['metadata']
+            }).get('metadata', {})
```
Comment (dstftw) [inline youtube_dl/extractor/archiveorg.py]: Point of this?

Comment (TinyToweringTree) [reply youtube_dl/extractor/archiveorg.py]: The idea is to make the extraction more resilient against website changes. With metadata being optional for downloading videos, it's probably better not to crash if the metadata element changes its name.

File: youtube_dl/extractor/archiveorg.py (hunk 1 above)
Comment (TinyToweringTree) [reply youtube_dl/extractor/archiveorg.py]: Thanks for the pointer. I changed the code to use extract_attributes() in commit e910f49.

It was still bugging me that I couldn't use get_element_by_class() in the line before as it only returns the *content* of the element and not the element (as the name of the function indicates). So I changed get_element_by_class() (and it's associated functions) in commit b98d1c0 to accept an optional `include_tag` parameter. By default it's set to `False`. So no existing code will be affected by it.

This should prove useful in the future and make those functions more intuitive to use.

I also added tests to `test_utils.py`. While doing so I noticed that get_element_by_class() didn't work for class names starting with a hyphen. I fixed that, too.

File: youtube_dl/extractor/archiveorg.py (hunk 2 above)
Comment (dstftw) [reply youtube_dl/extractor/archiveorg.py]: This is not enough. If you reconsider this optional then you must also:
1. Handle non dict metadata scenario.
2. Handle download failure.

Comment (TinyToweringTree) [reply youtube_dl/extractor/archiveorg.py]: You are right. Thanks. Commit 1326a5a fixes this by setting `fatal=False` when calling `_download_json()` and by checking the return value.

**Comment by lukaarma** [comment]
@dstftw this pull request needs more work?

**Pull request closed without merging**